jax/_src/state/discharge.py

```diff
@@ -241,3 +251,473 @@ def _closed_call_discharge_rule(
+  # for non-res stuff:
+  #   getting and setting => (nonzero ct, UndefinedPrimal arg)
+  #   just setting =>        (nonzero ct, not UndefinedPrimal, dummy value)
+  #   just getting =>        (zero ct   , UndefinedPrimal arg)
+  # for res stuff:
+  #                          (zero ct   , not UndefinedPrimal)
+  assert any(which_linear)
+  transpose_args = []
+  for x, ct in zip(args, in_cts):
+    if   type(ct) is     ad_util.Zero and not ad.is_undefined_primal(x):
+      # this is a residual, take x!
+      transpose_args.append(x)
+    elif type(ct) is     ad_util.Zero and     ad.is_undefined_primal(x):
+      # the loop was 'just getting', plug in a zero
+      transpose_args.append(ad_util.zeros_like_aval(x.aval))
+    elif type(ct) is not ad_util.Zero and not ad.is_undefined_primal(x):
+      # the loop was 'just setting', grab that cotangent! x is dummy
+      transpose_args.append(ct)
+    elif type(ct) is not ad_util.Zero and     ad.is_undefined_primal(x):
+      # the loop was 'getting and setting', grab that cotangent!
+      transpose_args.append(ct)
+  jaxpr_transpose_, consts = _transpose_jaxpr(jaxpr, which_linear)
+  jaxpr_transpose = hoist_consts_to_refs(jaxpr_transpose_)
+  which_linear = (*[False] * len(consts), *which_linear)
+  const_all_outs = run_state_p.bind(*consts, *transpose_args,
+                                    jaxpr=jaxpr_transpose,
+                                    which_linear=which_linear)
+  _, all_outs = split_list(const_all_outs, [len(consts)])
+  ct_outs = [ct if ad.is_undefined_primal(x) else None
+             for x, ct in zip(args, all_outs)]
+  return ct_outs
+ad.primitive_transposes[run_state_p] = _run_state_transpose
+
+@register_discharge_rule(run_state_p)
+def _run_state_discharge_rule(in_avals: Sequence[core.AbstractValue],
+                              out_avals: Sequence[core.AbstractValue],
+                              *args: Any, jaxpr: core.Jaxpr,
+                              which_linear: Sequence[bool]):
+  del out_avals
+  out_vals = run_state_p.bind(*args, jaxpr=jaxpr, which_linear=which_linear)
+  new_invals = []
+  for aval, out_val in zip(in_avals, out_vals):
+    new_invals.append(out_val if isinstance(aval, AbstractRef) else None)
+  return new_invals, out_vals
+
+def initial_style_jaxpr(
+    fun: Callable, in_tree: PyTreeDef, in_avals: Sequence[core.AbstractValue]
+  ) -> Tuple[core.Jaxpr, List[Any], PyTreeDef]:
+  return _initial_style_jaxpr(fun, in_tree, tuple(in_avals))
+
+@weakref_lru_cache
+def _initial_style_jaxpr(fun, in_tree, in_avals):
+  fun_, out_tree_thunk = api_util.flatten_fun_nokwargs(lu.wrap_init(fun), in_tree)
+  debug = pe.debug_info(fun_, in_tree, False, out_tree_thunk, 'run_state')
+  jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(fun_, in_avals, debug)
+  return jaxpr, consts, out_tree_thunk()
+
+def run_state(f: Callable[..., None]):
+  def wrapped(*args):
+    singleton = len(args) == 1
```

Comment:
This feels weird to me... as we discussed!

Reply:
Two options that seem more internally coherent:

```python
# "no splatting" option, not as nice for f but nicer for caller of run_state
def f(refs):
  x_ref, = refs
  ...

x_final = run_state(f)(x_init)
assert tree_structure(x_final) == tree_structure(x_init)
```

```python
# splat but tuple output, nicer for f but not as nice for caller of run_state (must unpack output)
def f(x_ref):
  ...

x_final, = run_state(f)(x_init)
assert tree_structure(x_final) == tree_structure(x_init)
```

Or there's the current option of special-casing single input to be automatically unpacked.

Reply:
Anything but special-casing, please. IMO, we already have one too many special-casing (`primitive.multiple_results`).

Reply:
Someday we're going to get rid of that! 😅

Reply:
I went w/ option 1